pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/UnnecessaryImportRule.java

```diff
@@ -230,12 +237,36 @@ public Object visit(ASTClassType node, Object data) {
         return super.visit(node, data);
     }
 
+    @Override
+    public Object visit(ASTAmbiguousName node, Object data) {
+        // ambiguous name means the symbol table could not resolve the first name
+
+        // only consider static imports
+        boolean onlyStatic = !(node.getParent() instanceof ASTClassType);
+        recordFailedTypeResWithName(node, node.getFirstToken().getImage(), onlyStatic);
+        return null;
+    }
+
+    private void recordFailedTypeResWithName(JavaNode location, String name, boolean onlyStatics) {
+        String target = onlyStatics ? "static " : "";
+        LOG.debug("UnnecessaryImport: Failed type res for {} will cause all {}imports named {} to be marked as used", location, target, name);
+        boolean foundNamedImport = allSingleNameImports.removeIf(
+            decl -> (!onlyStatics || decl.isStatic())
+                && name.equals(decl.node.getImportedSimpleName()));
+        if (!foundNamedImport) {
+            LOG.debug("+ Since no such named import can be found, all {}on-demand-imports will be marked as used", target);
+
+            allImportsOnDemand.removeIf(it -> !onlyStatics || it.isStatic());
+        }
+    }
```

Comment:
As we discussed in #3901, we need to work out a better way for users to detect and fix these incomplete auxclasspath issues that don't require them proactively running PMD in verbose mode and manually reading the massive non-structured logs to find them…